compiler/damlc/daml-stdlib-src/DA/Numeric.daml

```diff
@@ -54,3 +66,184 @@ shift = primitive @"BEShiftNumeric" (numericOne @n2)
 -- | The number pi.
 pi : forall n. NumericScale n => Numeric n
 pi = castAndRound (3.14159_26535_89793_23846_26433_83279_50288_41 : Numeric 37)
+
+zero37 : Numeric 37
+zero37 = 0.0
+
+one0 : Numeric 0
+one0 = 1.0
+
+delta37 : Numeric 37
+delta37 = shift one0
+
+roundHalfEven : forall n. NumericScale n => Numeric 37 -> Numeric n
+roundHalfEven x = castAndRound @n x
+
+roundCeiling : NumericScale n => Numeric 37 -> Numeric n
+roundCeiling x =
+   if (x <= cast r) then
+    r
+  else
+    r + shift one0
+  where
+    r = roundHalfEven x
+
+roundFloor : NumericScale n => Numeric 37 -> Numeric n
+roundFloor x =
+    if (cast r <= x) then
+      r
+    else
+      r - shift one0
+    where
+      r = roundHalfEven x
+
+roundUp : NumericScale n => Numeric 37 -> Numeric n
+roundUp x =
+    if (zero37 <= x) then
+      roundCeiling x
+    else
+      roundFloor x
+
+roundDown : NumericScale n => Numeric 37 -> Numeric n
+roundDown x =
+    if (zero37 <= x) then
+      roundFloor x
+    else
+      roundCeiling x
+
+roundHalfCeiling : NumericScale n => Numeric 37 -> Numeric n
+roundHalfCeiling x =
+    if (x <= cast r) then
+      r
+    else
+      roundHalfEven (x + delta37)
+    where
+      r = roundHalfEven x
+
+roundHalfFloor : NumericScale n => Numeric 37 -> Numeric n
+roundHalfFloor x =
+    if (cast r <= x) then
+      r
+    else
+      roundHalfEven (x - delta37)
+    where
+      r = roundHalfEven x
+
+roundHalfUp : NumericScale n => Numeric 37 -> Numeric n
+roundHalfUp x =
+    if (zero37 <= x) then
+     roundHalfCeiling x
+    else
+      roundHalfFloor x
+
+roundHalfDown : NumericScale n => Numeric 37 -> Numeric n
+roundHalfDown x =
+  if (zero37 <= x) then
+    roundHalfFloor x
+  else
+    roundHalfCeiling x
+
+roundUnnecessary : NumericScale n => Numeric 37 -> Numeric n
+roundUnnecessary = cast
+
+#ifndef DAML_BIGNUMERIC
+-- | Rounding modes.
+data RoundingMode
+  = RoundingUp -- ^ Round away from zero37.
+  | RoundingDown -- ^ Round towards zero37.
+  | RoundingCeiling -- ^ Round towards positive infinity.
+  | RoundingFloor -- ^ Round towards negative infinity.
+  | RoundingHalfUp
+      -- ^ Round towards the nearest neighbor unless both neighbors
+      -- are equidistant, in which case round away from zero37.
+  | RoundingHalfDown
+      -- ^ Round towards the nearest neighbor unless both neighbors
+      -- are equidistant, in which case round towards zero37.
+  | RoundingHalfEven
+      -- ^ Round towards the nearest neighbor unless both neighbors
+      -- are equidistant, in which case round towards the even neighbor.
+  | RoundingUnnecessary
```

Comment:
Is this the only variant that can cause an ArithmeticError? If so, maybe it would be best to move it out of the variant, so that people can use safe code only, if they want to.

Reply:
All the other variants can throw an `ArithmeticError` in case of overflow.